zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/identity/IdentityTenantService.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.impl.identity;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.camunda.identity.sdk.Identity;
+import io.camunda.identity.sdk.tenants.dto.Tenant;
+import io.camunda.zeebe.gateway.impl.configuration.IdentityRequestCfg;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+public class IdentityTenantService {
+
+  private static final TimeUnit TIME_UNIT = TimeUnit.MILLISECONDS;
+
+  private final LoadingCache<String, List<Tenant>> tenantCache;
+  private final Semaphore semaphore;
+  private final Identity identity;
+
+  private final boolean isCachingEnabled;
+  private final long semaphoreTimeout;
+  private final RejectedExecutionException ree;
+
+  public IdentityTenantService(final Identity identity, final IdentityRequestCfg config) {
+    this.identity = identity;
+    isCachingEnabled = config.isEnabled();
+    semaphoreTimeout = config.getTenantRequestTimeout();
+    semaphore = new Semaphore(config.getTenantRequestCapacity());
+    tenantCache =
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(config.getTenantCacheTtl(), TIME_UNIT)
+            .maximumSize(config.getTenantCacheSize())
+            .build(
+                new CacheLoader<>() {
+                  @Override
+                  public List<Tenant> load(final String token) {
+                    return getTenantsForTokenThrottled(token);
+                  }
+                });
+    ree =
+        new RejectedExecutionException(
```

Comment:
:question: So the stack trace will always show this right? We don't know where exactly it was thrown. I guess it's OK; not ideal, but it's definitely cheaper all things considered :thinking:

Reply:
@npepinpe considering your comment [here](https://github.com/camunda/camunda/pull/24196#discussion_r1822944741), I'll only throw it if the semaphore lock can't be acquired (on line 67). Does that make sense?

Reply:
UPDATE: I also created a new Zeebe `ConcurrentRequestException`, so it's something unique that can be mapped to this feature.